File: backend/app/restriction.py
"""Government / EMSS tagging and photography rules for the layout hierarchy."""

from __future__ import annotations

from typing import Any

from sqlalchemy.orm import Session

from app.models import AisleRow, Area, Device, Project, Rack, VisionSession

RESTRICTION_TYPES = ("government", "EMSS", "other")


def sync_flags(
    restricted: bool = False,
    restriction_type: str = "",
    photography_allowed: bool = True,
) -> dict[str, Any]:
    kind = (restriction_type or "").strip()
    if kind:
        return {"restricted": True, "restriction_type": kind, "photography_allowed": False}
    if restricted:
        return {"restricted": True, "restriction_type": "", "photography_allowed": False}
    return {
        "restricted": False,
        "restriction_type": "",
        "photography_allowed": bool(photography_allowed),
    }


def apply_flags(data: dict[str, Any]) -> dict[str, Any]:
    if not {"restricted", "restriction_type", "photography_allowed"} & set(data):
        return data
    data.update(
        sync_flags(
            restricted=bool(data.get("restricted")),
            restriction_type=str(data.get("restriction_type") or ""),
            photography_allowed=data.get("photography_allowed", True),
        )
    )
    return data


def flag_reasons(
    kind: str,
    name: str,
    *,
    restricted: bool = False,
    restriction_type: str = "",
    photography_allowed: bool = True,
    restricted_reason: str = "",
) -> list[str]:
    reasons: list[str] = []
    label = restriction_type or restricted_reason or "restricted"
    if restricted:
        reasons.append(f"{kind} {name} is flagged {label}")
    if photography_allowed is False:
        msg = f"Photography is forbidden on {kind} {name}"
        if msg not in reasons:
            reasons.append(msg)
    return reasons


def entity_reasons(kind: str, entity: Any | None) -> list[str]:
    if entity is None:
        return []
    name = getattr(entity, "name", "") or kind.lower()
    restricted = bool(getattr(entity, "restricted", False))
    restriction_type = str(getattr(entity, "restriction_type", "") or getattr(entity, "restricted_reason", "") or "")
    photography_allowed = getattr(entity, "photography_allowed", not restricted)
    if restricted and getattr(entity, "photography_allowed", None) is None:
        photography_allowed = False
    return flag_reasons(
        kind,
        name,
        restricted=restricted,
        restriction_type=restriction_type,
        photography_allowed=bool(photography_allowed),
        restricted_reason=str(getattr(entity, "restricted_reason", "") or ""),
    )


def resolve_session_location(
    db: Session, session: VisionSession
) -> tuple[Project | None, Area | None, AisleRow | None, Rack | None]:
    project = db.get(Project, session.project_id)
    rack = db.get(Rack, session.rack_id) if session.rack_id else None
    row = db.get(AisleRow, session.row_id) if session.row_id else None
    if not row and rack and rack.row_id:
        row = db.get(AisleRow, rack.row_id)
    area = db.get(Area, session.area_id) if session.area_id else None
    if not area and rack and rack.area_id:
        area = db.get(Area, rack.area_id)
    if not area and row and row.area_id:
        area = db.get(Area, row.area_id)
    return project, area, row, rack


def hierarchy_reasons(db: Session, session: VisionSession) -> list[str]:
    project, area, row, rack = resolve_session_location(db, session)
    reasons: list[str] = []
    reasons.extend(entity_reasons("Project", project))
    if rack:
        reasons.extend(entity_reasons("Rack", rack))
        reasons.extend(entity_reasons("Row", row))
    elif row:
        reasons.extend(entity_reasons("Row", row))
    elif area:
        reasons.extend(entity_reasons("Area", area))
    return reasons

File: frontend/src/restriction.ts
import type { AisleRow, Area, Project, Rack } from "./api";

export const RESTRICTION_TYPES = ["government", "EMSS", "other"] as const;
export type RestrictionType = "" | (typeof RESTRICTION_TYPES)[number];

export type Restrictable = {
  restricted?: boolean;
  restriction_type?: string;
  restricted_reason?: string;
  photography_allowed?: boolean;
};

export type RestrictionHit = { label: string; type: string };

export function restrictionFields(type: RestrictionType) {
  return {
    restricted: Boolean(type),
    restriction_type: type,
    photography_allowed: !type,
  };
}

export function deviceRestrictionFields(type: RestrictionType) {
  return {
    restricted: Boolean(type),
    restricted_reason: type,
  };
}

export function restrictionTypeOf(entity?: Restrictable | null): RestrictionType {
  if (!entity) return "";
  const raw = (entity.restriction_type || entity.restricted_reason || "").trim();
  if (raw === "government" || raw === "EMSS" || raw === "other") return raw;
  if (entity.restricted || entity.photography_allowed === false) return "other";
  return "";
}

export function isRestrictedEntity(entity?: Restrictable | null): boolean {
  if (!entity) return false;
  return Boolean(entity.restricted) || entity.photography_allowed === false;
}

export function inheritedPhotoBlockers(opts: {
  project?: Project | null;
  row?: AisleRow | null;
  rack?: Rack | null;
}): RestrictionHit[] {
  const hits: RestrictionHit[] = [];
  const push = (entity: Restrictable | null | undefined, label: string) => {
    if (!entity || !isRestrictedEntity(entity)) return;
    hits.push({ label, type: restrictionTypeOf(entity) || "restricted" });
  };
  push(opts.project, "project");
  if (opts.row) push(opts.row, `row ${opts.row.name}`);
  if (opts.rack) push(opts.rack, `rack ${opts.rack.name}`);
  return hits;
}

export function photosAllowed(opts: {
  project?: Project | null;
  area?: Area | null;
  row?: AisleRow | null;
  rack?: Rack | null;
  device?: Restrictable | null;
}): boolean {
  if (opts.device && isRestrictedEntity(opts.device)) return false;
  if (opts.rack && isRestrictedEntity(opts.rack)) return false;
  if (opts.row && isRestrictedEntity(opts.row)) return false;
  if (opts.project && isRestrictedEntity(opts.project)) return false;
  if (!opts.row && !opts.rack && !opts.device && opts.area && isRestrictedEntity(opts.area)) return false;
  return true;
}

export function restrictionCaption(entity?: Restrictable | null): string {
  const type = restrictionTypeOf(entity);
  if (type) return `${type} · no photos`;
  if (entity?.restricted) return "restricted · no photos";
  if (entity?.photography_allowed === false) return "photos forbidden";
  return "photos OK";
}
